File: leetcode_2022/udemy_leetcode/matrix_dfs.py
def find_dfs(self, matrix):
        ROWS = len(matrix)
        COLS = len(matrix[0])
        visited = set()
        result = []

        def dfs(row, col, visited):
            visited.add((row, col))
            result.append(matrix[row][col])
            dirs = [[-1, 0], [0, 1], [1, 0], [0, -1]]
            for dir in dirs:
                r = row + dir[0]
                c = col + dir[1]
                if r < 0 or r >= ROWS or c < 0 or c >= COLS or (r, c) in visited:
                    continue
                dfs(r, c, visited)

        dfs(0, 0, visited)
        return result

File: leetcode_2022/udemy_leetcode/matrix_dfs.js
const directions = [[-1,0],  //up
                    [0,1],  //right
                    [1,0],  //down
                    [0,-1]  //left
                    ]

const matrix =  [
        [1,2,3,4,5],
        [6,7,8,9,10],
        [11,12,13,14,15],
        [16,17,18,19,20]
    ]

const traversalDFS = (matrix) => {
    const seen = new Array(matrix.length).fill(0)
                .map(() => new Array(matrix[0].length).fill(false))

    const values = []

    dfs(matrix, 0, 0, seen, values)

    return values
}

const dfs = (matrix, row, col, seen, values) => {
    if(row < 0 || col < 0 || row >= matrix.length || col >= matrix[0].length || seen[row][col]) return;

    values.push(matrix[row][col])
    seen[row][col] = true    

    for(let i = 0; i< directions.length; i++){
        const currentDir = directions[i]

        dfs(matrix, row + currentDir[0], col+currentDir[1], seen, values)
    }
}



